In [1]:
# ============================================================
# SpaceX Launch Capacity & Starship Transition Intelligence
# Notebook 02: Data Cleaning and Validation
# ============================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
from datetime import datetime, timezone
import json

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
pd.set_option("display.width", 120)
sns.set_style("whitegrid")

print("Notebook 02 – Data Cleaning and Validation")
print("Philosophy: Observe carefully. Validate before deriving. Never overclaim.")

Notebook 02 – Data Cleaning and Validation
Philosophy: Observe carefully. Validate before deriving. Never overclaim.


In [2]:
PROJECT_ROOT = Path.cwd().parent

RAW_DIR = PROJECT_ROOT / "data" / "raw" / "launches"
INTERIM_DIR = PROJECT_ROOT / "data" / "interim"
PROCESSED_DIR = PROJECT_ROOT / "data" / "processed"

INTERIM_DIR.mkdir(parents=True, exist_ok=True)
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

print("Paths configured")
print(f"Raw data location : {RAW_DIR}")
print(f"Interim data location : {INTERIM_DIR}")
print(f"Processed data location : {PROCESSED_DIR}")

Paths configured
Raw data location : c:\Users\Administrator\Desktop\Space--X--Launch-Intelligence\data\raw\launches
Interim data location : c:\Users\Administrator\Desktop\Space--X--Launch-Intelligence\data\interim
Processed data location : c:\Users\Administrator\Desktop\Space--X--Launch-Intelligence\data\processed


In [3]:
raw_path = RAW_DIR / "falcon9_launches_raw.csv"

df = pd.read_csv(raw_path)

print("Raw data loaded successfully")
print(f"Records : {len(df):,}")
print(f"Columns : {len(df.columns)}")

Raw data loaded successfully
Records : 819
Columns : 348


In [4]:
print("=== DATA PROVENANCE ===")
print(f"Source file          : {raw_path.name}")
print(f"Records loaded       : {len(df):,}")

if "collection_timestamp" in df.columns:
    print(f"Collection timestamp : {df['collection_timestamp'].iloc[0]}")
if "api_version" in df.columns:
    print(f"API version          : {df['api_version'].iloc[0]}")
if "data_source" in df.columns:
    print(f"Data source          : {df['data_source'].iloc[0]}")

=== DATA PROVENANCE ===
Source file          : falcon9_launches_raw.csv
Records loaded       : 819
Collection timestamp : 2026-08-19T08:03:45.329706+00:00
API version          : 2.3.0
Data source          : Launch Library 2


In [5]:
print("=== INITIAL INSPECTION ===")
print("\nShape:", df.shape)
print("\nMissing values (Top 15):")
print(df.isna().sum().sort_values(ascending=False).head(15))

=== INITIAL INSPECTION ===

Shape: (819, 348)

Missing values (Top 15):
net_precision                                    819
mission.image                                    819
rocket.configuration.manufacturer.image          819
pad.image                                        819
pad.info_url                                     819
rocket.configuration.geo_capacity                819
rocket.configuration.alias                       819
rocket.configuration.manufacturer.parent         819
mission.orbit                                    819
rocket.configuration.manufacturer.social_logo    819
launch_service_provider.parent                   818
rocket.configuration.sso_capacity                817
failreason                                       815
hashtag                                          798
infographic                                      771
dtype: int64


In [6]:
print("=== DUPLICATE CHECK ===")

duplicate_count = df["id"].duplicated().sum()
print(f"Duplicate launch IDs: {duplicate_count}")

if duplicate_count > 0:
    print("\nDuplicate records found:")
    display(df[df["id"].duplicated(keep=False)].sort_values("id"))
else:
    print("No duplicate launch IDs found.")

=== DUPLICATE CHECK ===
Duplicate launch IDs: 8

Duplicate records found:


,id,url,name,response_mode,slug,launch_designator,last_updated,net,net_precision,window_end,window_start,infographic,probability,weather_concerns,failreason,hashtag,webcast_live,program,orbital_launch_attempt_count,location_launch_attempt_count,pad_launch_attempt_count,agency_launch_attempt_count,orbital_launch_attempt_count_year,location_launch_attempt_count_year,pad_launch_attempt_count_year,agency_launch_attempt_count_year,flightclub_url,updates,info_urls,vid_urls,timeline,pad_turnaround,mission_patches,status.id,status.name,status.abbrev,status.description,image.id,image.name,image.image_url,image.thumbnail_url,image.credit,image.license.id,image.license.name,image.license.priority,image.license.link,image.single_use,image.variants,launch_service_provider.response_mode,launch_service_provider.id,launch_service_provider.url,launch_service_provider.name,launch_service_provider.abbrev,launch_service_provider.type.id,launch_service_provider.type.name,launch_service_provider.featured,launch_service_provider.country,launch_service_provider.description,launch_service_provider.administrator,launch_service_provider.founding_year,launch_service_provider.launchers,launch_service_provider.spacecraft,launch_service_provider.parent,launch_service_provider.image.id,launch_service_provider.image.name,launch_service_provider.image.image_url,launch_service_provider.image.thumbnail_url,launch_service_provider.image.credit,launch_service_provider.image.license.id,launch_service_provider.image.license.name,launch_service_provider.image.license.priority,launch_service_provider.image.license.link,launch_service_provider.image.single_use,launch_service_provider.image.variants,launch_service_provider.logo.id,launch_service_provider.logo.name,launch_service_provider.logo.image_url,launch_service_provider.logo.thumbnail_url,launch_service_provider.logo.credit,launch_service_provider.logo.license.id,launch_service_provider.logo.license.name,launch_service_provider.logo.license.priority,launch_service_provider.logo.license.link,launch_service_provider.logo.single_use,launch_service_provider.logo.variants,launch_service_provider.social_logo.id,launch_service_provider.social_logo.name,launch_service_provider.social_logo.image_url,launch_service_provider.social_logo.thumbnail_url,launch_service_provider.social_logo.credit,launch_service_provider.social_logo.license.id,launch_service_provider.social_logo.license.name,launch_service_provider.social_logo.license.priority,launch_service_provider.social_logo.license.link,launch_service_provider.social_logo.single_use,launch_service_provider.social_logo.variants,launch_service_provider.total_launch_count,launch_service_provider.consecutive_successful_launches,launch_service_provider.successful_launches,launch_service_provider.failed_launches,launch_service_provider.pending_launches,launch_service_provider.consecutive_successful_landings,launch_service_provider.successful_landings,launch_service_provider.failed_landings,launch_service_provider.attempted_landings,launch_service_provider.successful_landings_spacecraft,launch_service_provider.failed_landings_spacecraft,launch_service_provider.attempted_landings_spacecraft,launch_service_provider.successful_landings_payload,launch_service_provider.failed_landings_payload,launch_service_provider.attempted_landings_payload,launch_service_provider.info_url,launch_service_provider.wiki_url,launch_service_provider.social_media_links,rocket.id,rocket.configuration.response_mode,rocket.configuration.id,rocket.configuration.url,rocket.configuration.name,rocket.configuration.families,rocket.configuration.full_name,rocket.configuration.variant,rocket.configuration.active,rocket.configuration.is_placeholder,rocket.configuration.manufacturer.response_mode,rocket.configuration.manufacturer.id,rocket.configuration.manufacturer.url,rocket.configuration.manufacturer.name,rocket.configuration.manufacturer.abbrev,rocket.configuration.manufacturer.type.id,rocket.configuration.manufac

In [7]:
print("=== FALCON 9 CANDIDATE FILTERING ===")

df["vehicle_name"] = (
    df["rocket.configuration.full_name"]
    .astype("string")
    .str.strip()
)

df["is_falcon9_candidate"] = (
    df["vehicle_name"]
    .str.contains("Falcon 9", case=False, na=False)
)

df_f9 = df[df["is_falcon9_candidate"]].copy()
excluded_records = df[~df["is_falcon9_candidate"]].copy()

print(f"Falcon 9 candidates : {len(df_f9):,}")
print(f"Excluded records    : {len(excluded_records):,}")

print("\nVehicle name distribution (candidates):")
print(df_f9["vehicle_name"].value_counts())

=== FALCON 9 CANDIDATE FILTERING ===
Falcon 9 candidates : 789
Excluded records    : 30

Vehicle name distribution (candidates):
vehicle_name
Falcon 9 Block 5        732
Falcon 9 Full Thrust     25
Falcon 9 v1.1            15
Falcon 9 Block 4         12
Falcon 9 v1.0             5
Name: count, dtype: int64[pyarrow]


C:\Users\Administrator\AppData\Local\Temp\ipykernel_11336\2071299023.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["vehicle_name"] = (
C:\Users\Administrator\AppData\Local\Temp\ipykernel_11336\2071299023.py:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df["is_falcon9_candidate"] = (


In [8]:
print("=== EXTRACT CORE FIELDS ===")

core_mapping = {
    "id": "launch_id",
    "name": "mission_name",
    "net": "launch_date_raw",
    "status.name": "status",
    "vehicle_name": "vehicle",
    "pad.name": "launch_pad",
    "pad.location.name": "launch_site"
}

available_cols = [col for col in core_mapping.keys() if col in df_f9.columns]
df_clean = df_f9[available_cols].rename(columns=core_mapping)

print("Core fields extracted")
print(f"Working dataset shape: {df_clean.shape}")
df_clean.head(3)

=== EXTRACT CORE FIELDS ===
Core fields extracted
Working dataset shape: (789, 7)


,launch_id,mission_name,launch_date_raw,status,vehicle,launch_pad,launch_site
0,a5556e79-a01f-48c6-b822-69bbadb4460d,Falcon 9 Block 5 | Crew-14,2031-03-31T00:00:00Z,To Be Determined,Falcon 9 Block 5,Unknown Pad,"Cape Canaveral SFS, FL, USA"
1,c277ebf9-3ce3-49d6-b212-2f350a2b89ee,Falcon 9 Block 5 | Amazon Leo (LF-05),2029-12-31T00:00:00Z,To Be Determined,Falcon 9 Block 5,Unknown Pad,"Cape Canaveral SFS, FL, USA"
3,3930a4f9-1bd4-4f83-8906-70fcda1e31cd,Falcon 9 Block 5 | Amazon Leo (LF-10),2029-12-31T00:00:00Z,To Be Determined,Falcon 9 Block 5,Unknown Pad,"Cape Canaveral SFS, FL, USA"


In [9]:
print("=== DATE VALIDATION ===")

df_clean["launch_date"] = pd.to_datetime(
    df_clean["launch_date_raw"],
    errors="coerce",
    utc=True
)

invalid_dates = df_clean["launch_date"].isna().sum()
print(f"Invalid / missing dates: {invalid_dates}")

print(f"Earliest launch : {df_clean['launch_date'].min()}")
print(f"Latest launch   : {df_clean['launch_date'].max()}")

=== DATE VALIDATION ===
Invalid / missing dates: 0
Earliest launch : 2010-06-04 18:45:00+00:00
Latest launch   : 2031-03-31 00:00:00+00:00


In [10]:
print("=== POTENTIAL BOOSTER-RELATED COLUMNS ===")

booster_related = [
    col for col in df_f9.columns
    if any(keyword in col.lower() for keyword in 
           ["launcher", "serial", "core", "booster", "flight"])
]

if booster_related:
    for col in booster_related:
        print(f"  - {col}")
else:
    print("No obvious booster-related columns found in this extract.")

=== POTENTIAL BOOSTER-RELATED COLUMNS ===
  - flightclub_url
  - launch_service_provider.launchers
  - rocket.configuration.manufacturer.launchers
  - rocket.configuration.maiden_flight
  - rocket.launcher_stage


In [11]:
print("=== EXTRACT BOOSTER INFORMATION (FROM LAUNCH RECORDS) ===")

def first_valid_value(row, fields):
    for field in fields:
        if field in row.index:
            value = row[field]
            if pd.notna(value):
                value = str(value).strip()
                if value and value.lower() not in ["nan", "none", "unknown", ""]:
                    return value
    return pd.NA

serial_fields = [
    "rocket.first_stage.0.launcher.serial_number",
    "rocket.launcher_stage.0.launcher.serial_number",
    "rocket.launcher_stage.0.serial_number",
]

flight_fields = [
    "rocket.first_stage.0.launcher.flights",
    "rocket.launcher_stage.0.launcher.flights",
    "rocket.launcher_stage.0.flight",
]

df_f9["booster_serial"] = df_f9.apply(lambda row: first_valid_value(row, serial_fields), axis=1)
df_f9["booster_flight_number"] = df_f9.apply(lambda row: first_valid_value(row, flight_fields), axis=1)

print("Booster serial available      :", df_f9["booster_serial"].notna().sum())
print("Booster flight number available:", df_f9["booster_flight_number"].notna().sum())

display(df_f9[["name", "booster_serial", "booster_flight_number"]].head(15))

=== EXTRACT BOOSTER INFORMATION (FROM LAUNCH RECORDS) ===
Booster serial available      : 0
Booster flight number available: 0


,name,booster_serial,booster_flight_number
0,Falcon 9 Block 5 | Crew-14,<NA>,<NA>
1,Falcon 9 Block 5 | Amazon Leo (LF-05),<NA>,<NA>
3,Falcon 9 Block 5 | Amazon Leo (LF-10),<NA>,<NA>
4,Falcon 9 Block 5 | Amazon Leo (LF-13),<NA>,<NA>
5,Falcon 9 Block 5 | JSAT-31,<NA>,<NA>
7,Falcon 9 Block 5 | Amazon Leo (LF-08),<NA>,<NA>
8,Falcon 9 Block 5 | NROL-96,<NA>,<NA>
9,Falcon 9 Block 5 | Amazon Leo (LF-09),<NA>,<NA>
10,Falcon 9 Block 5 | Amazon Leo (LF-06),<NA>,<NA>
11,Falcon 9 Block 5 | Amazon Leo (LF-07),<NA>,<NA>


In [12]:
print("=== BOOSTER DATA COVERAGE ===")

total_f9 = len(df_f9)
serial_count = df_f9["booster_serial"].notna().sum()
flight_count = df_f9["booster_flight_number"].notna().sum()

coverage_report = pd.DataFrame({
    "Metric": [
        "Total Falcon 9 records",
        "Booster serial available",
        "Booster flight number available",
        "Serial coverage %",
        "Flight number coverage %"
    ],
    "Value": [
        total_f9,
        serial_count,
        flight_count,
        round(serial_count / total_f9 * 100, 2) if total_f9 > 0 else 0,
        round(flight_count / total_f9 * 100, 2) if total_f9 > 0 else 0
    ]
})

display(coverage_report)

=== BOOSTER DATA COVERAGE ===


,Metric,Value
0,Total Falcon 9 records,789.0
1,Booster serial available,0.0
2,Booster flight number available,0.0
3,Serial coverage %,0.0
4,Flight number coverage %,0.0


In [13]:
# ============================================================
# VALIDATE BOOSTER SERIAL VALUES
# ============================================================

invalid_values = ["unknown", "n/a", "na", "none", "null", ""]

df_f9["booster_serial"] = (
    df_f9["booster_serial"]
    .astype("string")
    .str.strip()
)

df_f9.loc[
    df_f9["booster_serial"].str.lower().isin(invalid_values),
    "booster_serial"
] = pd.NA

print("Valid booster serials after cleaning:", df_f9["booster_serial"].notna().sum())

Valid booster serials after cleaning: 0


In [14]:
print("=== PRELIMINARY MISSION CLASSIFICATION ===")

def classify_mission(name):
    if pd.isna(name):
        return "Unknown"
    
    name = str(name).lower()
    
    if "starlink" in name:
        return "Starlink"
    elif any(x in name for x in ["crew-", "crs-", "axiom", "ax-", "inspiration", "polaris"]):
        return "Crew / Human Spaceflight"
    elif any(x in name for x in ["nrol", "ussf", "nro", "gps", "otv", "stp-"]):
        return "Government / National Security"
    elif any(x in name for x in ["transporter", "bandwagon"]):
        return "Rideshare"
    else:
        return "Commercial / Other"

df_f9["mission_type"] = df_f9["name"].apply(classify_mission)

print("\nMission Type Distribution (Preliminary):")
display(df_f9["mission_type"].value_counts(dropna=False).to_frame("Launches"))

print("\nNote: Mission type is preliminary and rule-based. It must be validated before final capacity analysis.")

=== PRELIMINARY MISSION CLASSIFICATION ===

Mission Type Distribution (Preliminary):


,Launches
mission_type,
Starlink,419
Commercial / Other,244
Crew / Human Spaceflight,62
Government / National Security,42
Rideshare,22



Note: Mission type is preliminary and rule-based. It must be validated before final capacity analysis.


In [15]:
print("=== PREPARE TURNAROUND DATASET ===")

df_boosters = df_f9[df_f9["booster_serial"].notna()].copy()
df_boosters["launch_date"] = pd.to_datetime(df_boosters["net"], errors="coerce", utc=True)

print("Records with booster serial :", len(df_boosters))
print("Records with valid launch date:", df_boosters["launch_date"].notna().sum())

df_boosters = df_boosters.dropna(subset=["launch_date"])
df_boosters = df_boosters.sort_values(["booster_serial", "launch_date"])

=== PREPARE TURNAROUND DATASET ===
Records with booster serial : 0
Records with valid launch date: 0


In [16]:
df_boosters["previous_launch_date"] = (
    df_boosters.groupby("booster_serial")["launch_date"].shift(1)
)

df_boosters["turnaround_days"] = (
    (df_boosters["launch_date"] - df_boosters["previous_launch_date"])
    .dt.total_seconds() / 86400
)

print("Turnaround calculation complete.")
print("Records with calculable turnaround:", df_boosters["turnaround_days"].notna().sum())

Turnaround calculation complete.
Records with calculable turnaround: 0


In [17]:
print("=== TURNAROUND VALIDATION ===")

negative = df_boosters[df_boosters["turnaround_days"] < 0]
zero = df_boosters[df_boosters["turnaround_days"] == 0]

print(f"Negative turnaround records : {len(negative)}")
print(f"Zero-day turnaround records : {len(zero)}")

if len(negative) > 0:
    display(negative[["mission_name", "booster_serial", "launch_date", "previous_launch_date", "turnaround_days"]])

=== TURNAROUND VALIDATION ===
Negative turnaround records : 0
Zero-day turnaround records : 0


In [19]:
print("=== FINAL DATA QUALITY REPORT ===")

# assemble final analysis-ready dataframe
df_final = df_clean.copy()

# bring in booster identifiers and mission type from df_f9
_booster_cols = df_f9[["id", "booster_serial", "booster_flight_number", "mission_type"]].rename(columns={"id": "launch_id"})
df_final = df_final.merge(_booster_cols, on="launch_id", how="left")

# bring in turnaround info from df_boosters (if available)
_turnaround = df_boosters[["id", "previous_launch_date", "turnaround_days"]].rename(columns={"id": "launch_id"})
df_final = df_final.merge(_turnaround, on="launch_id", how="left")

# ensure launch_date is datetime
df_final["launch_date"] = pd.to_datetime(df_final.get("launch_date"), utc=True, errors="coerce")

quality_report = pd.DataFrame({
    "column": df_final.columns,
    "dtype": df_final.dtypes.astype(str).values,
    "missing": df_final.isna().sum().values,
    "missing_pct": (df_final.isna().mean() * 100).round(2).values,
    "unique": df_final.nunique().values
})

display(quality_report)

# Save
interim_path = INTERIM_DIR / "falcon9_launches_with_boosters.csv"
processed_path = PROCESSED_DIR / "falcon9_analysis_ready.csv"

df_final.to_csv(interim_path, index=False)
df_final.to_csv(processed_path, index=False)

print(f"\nInterim saved  : {interim_path}")
print(f"Processed saved: {processed_path}")

=== FINAL DATA QUALITY REPORT ===


,column,dtype,missing,missing_pct,unique
0,launch_id,str,0,0.0,782
1,mission_name,str,0,0.0,782
2,launch_date_raw,str,0,0.0,705
3,status,str,0,0.0,5
4,vehicle,string,0,0.0,5
5,launch_pad,str,0,0.0,4
6,launch_site,str,0,0.0,3
7,launch_date,"datetime64[us, UTC]",0,0.0,705
8,booster_serial,string,803,100.0,0
9,booster_flight_number,object,803,100.0,0



Interim saved  : c:\Users\Administrator\Desktop\Space--X--Launch-Intelligence\data\interim\falcon9_launches_with_boosters.csv
Processed saved: c:\Users\Administrator\Desktop\Space--X--Launch-Intelligence\data\processed\falcon9_analysis_ready.csv


In [21]:
# ============================================================
# BOOSTER-LEVEL SUMMARY
# ============================================================

booster_summary = (
    df_final
    .groupby("booster_serial")
    .agg(
        observed_launches=("launch_id", "count"),
        first_observed_launch=("launch_date", "min"),
        last_observed_launch=("launch_date", "max"),
        max_recorded_flight=("booster_flight_number", "max"),
        median_turnaround_days=("turnaround_days", "median"),
        mean_turnaround_days=("turnaround_days", "mean")
    )
    .reset_index()
)

display(booster_summary.sort_values("max_recorded_flight", ascending=False).head(20))

booster_summary_path = PROCESSED_DIR / "falcon9_booster_summary.csv"
booster_summary.to_csv(booster_summary_path, index=False)

print(f"Booster summary saved → {booster_summary_path.name}")

,booster_serial,observed_launches,first_observed_launch,last_observed_launch,max_recorded_flight,median_turnaround_days,mean_turnaround_days


Booster summary saved → falcon9_booster_summary.csv


In [22]:
print("=== INSPECTING ROCKET / LAUNCHER STRUCTURE ===")

# Look at the first few Falcon 9 records more carefully
sample = df_f9.head(3)

# Check if rocket.launcher_stage exists and what it contains
if "rocket.launcher_stage" in df_f9.columns:
    print("\nrocket.launcher_stage sample:")
    print(sample["rocket.launcher_stage"].tolist())
else:
    print("\nColumn 'rocket.launcher_stage' does not exist")

# Check all columns that contain "launcher" or "serial"
launcher_cols = [col for col in df_f9.columns if "launcher" in col.lower() or "serial" in col.lower()]
print("\nAll columns containing 'launcher' or 'serial':")
for col in launcher_cols:
    print(f"  - {col}")

=== INSPECTING ROCKET / LAUNCHER STRUCTURE ===

rocket.launcher_stage sample:
['[]', '[]', '[]']

All columns containing 'launcher' or 'serial':
  - launch_service_provider.launchers
  - rocket.configuration.manufacturer.launchers
  - rocket.launcher_stage
  - booster_serial


In [23]:
# ============================================================
# COLLECT FALCON 9 BOOSTERS (LAUNCHERS)
# ============================================================

import requests
import json
import time
from datetime import datetime, timezone
from pathlib import Path

API_VERSION = "2.3.0"
LAUNCHERS_URL = f"https://ll.thespacedevs.com/{API_VERSION}/launchers/"

RAW_BOOSTERS_DIR = Path.cwd().parent / "data" / "raw" / "boosters"
RAW_BOOSTERS_DIR.mkdir(parents=True, exist_ok=True)

session = requests.Session()
session.headers.update({
    "User-Agent": "SpaceX-Launch-Intelligence/1.0 (Research Project)",
    "Accept": "application/json"
})

print("Launcher collection ready")
print(f"Endpoint: {LAUNCHERS_URL}")

Launcher collection ready
Endpoint: https://ll.thespacedevs.com/2.3.0/launchers/


In [24]:
all_launchers = []
url = LAUNCHERS_URL
params = {
    "limit": 100,
    "mode": "detailed",
    "ordering": "-flights"          # highest flight count first
}

total_available = None
page = 1

print("Starting launcher collection...\n")

while url:
    try:
        response = session.get(
            url,
            params=params if page == 1 else None,
            timeout=40
        )

        if response.status_code == 429:
            print("Rate limit reached. Stopping.")
            break

        response.raise_for_status()
        data = response.json()

        if total_available is None:
            total_available = data.get("count")
            print(f"API reports {total_available} launcher records\n")

        results = data.get("results", [])
        all_launchers.extend(results)

        # Save progress after every page
        with open(RAW_BOOSTERS_DIR / "launchers_raw.json", "w", encoding="utf-8") as f:
            json.dump(all_launchers, f)

        print(f"Page {page:02d} | Collected: {len(all_launchers):4d} launchers")

        url = data.get("next")
        page += 1
        time.sleep(2.5)

    except Exception as e:
        print(f"Error: {e}")
        break

print("\n=== Collection finished ===")
print(f"Total launchers collected: {len(all_launchers)}")

Starting launcher collection...

API reports 192 launcher records

Page 01 | Collected:  100 launchers
Page 02 | Collected:  192 launchers

=== Collection finished ===
Total launchers collected: 192


In [25]:
# Convert to DataFrame
df_launchers = pd.json_normalize(all_launchers)

print("Total launchers:", len(df_launchers))

# Keep only Falcon 9 related boosters
if "launcher_config.full_name" in df_launchers.columns:
    mask = df_launchers["launcher_config.full_name"].str.contains("Falcon 9", case=False, na=False)
elif "launcher_config.name" in df_launchers.columns:
    mask = df_launchers["launcher_config.name"].str.contains("Falcon 9", case=False, na=False)
else:
    mask = df_launchers["serial_number"].str.startswith("B", na=False)  # fallback

df_f9_boosters = df_launchers[mask].copy()

print(f"Falcon 9 boosters found: {len(df_f9_boosters)}")

# Save
df_f9_boosters.to_csv(RAW_BOOSTERS_DIR / "falcon9_boosters_raw.csv", index=False)

print("Saved → falcon9_boosters_raw.csv")
df_f9_boosters[["serial_number", "flights", "status.name"]].head(15)

Total launchers: 192
Falcon 9 boosters found: 87
Saved → falcon9_boosters_raw.csv


,serial_number,flights,status.name
3,Unknown F9,NaN,N/A
9,B1067,36.0,active
10,B1071,35.0,active
11,B1063,34.0,active
12,B1069,32.0,expended
13,B1077,30.0,active
14,B1078,29.0,active
15,B1080,28.0,active
16,B1081,26.0,active
17,B1062,23.0,destroyed


In [26]:
print("=== BOOSTER DATA STRUCTURE ===")
print("Columns available:")
print(df_f9_boosters.columns.tolist())

print("\nSample of key columns:")
display(df_f9_boosters[["serial_number", "flights", "status.name", "first_launch_date", "last_launch_date"]].head(10))

=== BOOSTER DATA STRUCTURE ===
Columns available:
['response_mode', 'id', 'url', 'flight_proven', 'serial_number', 'is_placeholder', 'details', 'successful_landings', 'attempted_landings', 'flights', 'last_launch_date', 'first_launch_date', 'fastest_turnaround', 'status.id', 'status.name', 'image.id', 'image.name', 'image.image_url', 'image.thumbnail_url', 'image.credit', 'image.license.id', 'image.license.name', 'image.license.priority', 'image.license.link', 'image.single_use', 'image.variants', 'launcher_config.response_mode', 'launcher_config.id', 'launcher_config.url', 'launcher_config.name', 'launcher_config.families', 'launcher_config.full_name', 'launcher_config.variant', 'image']

Sample of key columns:


,serial_number,flights,status.name,first_launch_date,last_launch_date
3,Unknown F9,NaN,N/A,NaN,NaN
9,B1067,36.0,active,2021-06-03T17:29:15Z,2026-07-09T09:25:43Z
10,B1071,35.0,active,2022-02-02T20:27:26Z,2026-07-11T03:01:57Z
11,B1063,34.0,active,2020-11-21T17:17:08Z,2026-08-04T17:05:03Z
12,B1069,32.0,expended,2021-12-21T10:07:08Z,2026-07-21T21:15:00Z
13,B1077,30.0,active,2022-10-05T16:00:57Z,2026-08-05T07:49:00Z
14,B1078,29.0,active,2023-03-02T05:34:14Z,2026-06-23T10:53:00Z
15,B1080,28.0,active,2023-05-21T21:37:09Z,2026-07-14T09:10:12Z
16,B1081,26.0,active,2023-08-26T07:27:27Z,2026-08-01T03:08:43Z
17,B1062,23.0,destroyed,2020-11-05T23:24:23Z,2024-08-28T07:48:00Z


In [27]:
# ============================================================
# CLEAN FALCON 9 BOOSTER DATASET
# ============================================================

# Keep only useful columns
booster_cols = {
    "id": "booster_id",
    "serial_number": "booster_serial",
    "flights": "total_flights",
    "status.name": "status",
    "first_launch_date": "first_launch_date",
    "last_launch_date": "last_launch_date",
    "successful_landings": "successful_landings",
    "attempted_landings": "attempted_landings",
    "fastest_turnaround": "fastest_turnaround",
    "launcher_config.full_name": "vehicle_config"
}

available = [c for c in booster_cols.keys() if c in df_f9_boosters.columns]
df_boosters_clean = df_f9_boosters[available].rename(columns=booster_cols)

# Remove placeholder / unknown boosters
df_boosters_clean = df_boosters_clean[
    ~df_boosters_clean["booster_serial"].str.contains("Unknown", case=False, na=False)
].copy()

# Convert dates
df_boosters_clean["first_launch_date"] = pd.to_datetime(df_boosters_clean["first_launch_date"], errors="coerce", utc=True)
df_boosters_clean["last_launch_date"] = pd.to_datetime(df_boosters_clean["last_launch_date"], errors="coerce", utc=True)

# Convert flight numbers
df_boosters_clean["total_flights"] = pd.to_numeric(df_boosters_clean["total_flights"], errors="coerce")

print("Clean booster dataset shape:", df_boosters_clean.shape)
print("\nTop boosters by flight count:")
display(df_boosters_clean.sort_values("total_flights", ascending=False).head(15))

Clean booster dataset shape: (86, 10)

Top boosters by flight count:


,booster_id,booster_serial,total_flights,status,first_launch_date,last_launch_date,successful_landings,attempted_landings,fastest_turnaround,vehicle_config
9,101,B1067,36.0,active,2021-06-03 17:29:15+00:00,2026-07-09 09:25:43+00:00,36.0,36.0,P22DT16H51M10S,Falcon 9 Block 5
10,107,B1071,35.0,active,2022-02-02 20:27:26+00:00,2026-07-11 03:01:57+00:00,35.0,35.0,P19DT16H19M25S,Falcon 9 Block 5
11,84,B1063,34.0,active,2020-11-21 17:17:08+00:00,2026-08-04 17:05:03+00:00,34.0,34.0,P32DT18H36M20S,Falcon 9 Block 5
12,105,B1069,32.0,expended,2021-12-21 10:07:08+00:00,2026-07-21 21:15:00+00:00,31.0,31.0,P25DT18H20M50S,Falcon 9 Block 5
13,116,B1077,30.0,active,2022-10-05 16:00:57+00:00,2026-08-05 07:49:00+00:00,30.0,30.0,P26DT20H34M40S,Falcon 9 Block 5
14,120,B1078,29.0,active,2023-03-02 05:34:14+00:00,2026-06-23 10:53:00+00:00,29.0,29.0,P20DT11H50M20S,Falcon 9 Block 5
15,123,B1080,28.0,active,2023-05-21 21:37:09+00:00,2026-07-14 09:10:12+00:00,28.0,28.0,P13DT12H34M20S,Falcon 9 Block 5
16,124,B1081,26.0,active,2023-08-26 07:27:27+00:00,2026-08-01 03:08:43+00:00,26.0,26.0,P24DT19H13M20S,Falcon 9 Block 5
17,73,B1062,23.0,destroyed,2020-11-05 23:24:23+00:00,2024-08-28 07:48:00+00:00,22.0,23.0,P21DT6H9M59S,Falcon 9 Block 5
18,125,B1082,23.0,active,2024-01-03 03:44:20+00:00,2026-07-21 14:49:34+00:00,23.0,23.0,P25DT10M,Falcon 9 Block 5


In [30]:
print("=== HANDLING DUPLICATE LAUNCH IDs ===")

# Keep the first occurrence of each launch_id
df_f9 = df_f9.drop_duplicates(subset=["id"], keep="first").copy()

print(f"Records after removing duplicates: {len(df_f9)}")
print(f"Remaining duplicate IDs: {df_f9['id'].duplicated().sum()}")

=== HANDLING DUPLICATE LAUNCH IDs ===
Records after removing duplicates: 782
Remaining duplicate IDs: 0


In [31]:
print("=== SEPARATING HISTORICAL vs FUTURE LAUNCHES ===")

df_f9["launch_date"] = pd.to_datetime(df_f9["net"], errors="coerce", utc=True)

today = pd.Timestamp.now(tz="UTC")

df_historical = df_f9[df_f9["launch_date"] <= today].copy()
df_future = df_f9[df_f9["launch_date"] > today].copy()

print(f"Historical / completed launches : {len(df_historical)}")
print(f"Future / scheduled launches     : {len(df_future)}")
print(f"\nHistorical date range:")
print(f"From: {df_historical['launch_date'].min()}")
print(f"To:   {df_historical['launch_date'].max()}")

=== SEPARATING HISTORICAL vs FUTURE LAUNCHES ===
Historical / completed launches : 680
Future / scheduled launches     : 102

Historical date range:
From: 2010-06-04 18:45:00+00:00
To:   2026-08-19 04:01:07+00:00


In [32]:
print("=== CREATE CLEAN LAUNCH DATASET ===")

final_cols = {
    "id": "launch_id",
    "name": "mission_name",
    "launch_date": "launch_date",
    "status.name": "status",
    "vehicle_name": "vehicle",
    "mission_type": "mission_type",
    "pad.name": "launch_pad",
    "pad.location.name": "launch_site"
}

available = [c for c in final_cols.keys() if c in df_historical.columns]
df_launches_clean = df_historical[available].rename(columns=final_cols)

print("Clean historical launch dataset shape:", df_launches_clean.shape)
display(df_launches_clean.head())

=== CREATE CLEAN LAUNCH DATASET ===
Clean historical launch dataset shape: (680, 8)


,launch_id,mission_name,launch_date,status,vehicle,mission_type,launch_pad,launch_site
124,f6958a77-37bb-4559-a108-40ba0ff9e34a,Falcon 9 Block 5 | Starlink Group 17-50,2026-08-19 04:01:07+00:00,Launch Successful,Falcon 9 Block 5,Starlink,Space Launch Complex 4E,"Vandenberg SFB, CA, USA"
125,55b79515-ddbf-4198-8a59-29a12ed21330,Falcon 9 Block 5 | USSF-366,2026-08-16 01:50:45+00:00,Launch Successful,Falcon 9 Block 5,Government / National Security,Space Launch Complex 4E,"Vandenberg SFB, CA, USA"
126,be8b4b0d-1989-48a5-a96d-e8e4d59ed8b3,Falcon 9 Block 5 | Globalstar 2-R Mission 1 (x 8),2026-08-16 01:12:14+00:00,Launch Successful,Falcon 9 Block 5,Commercial / Other,Space Launch Complex 40,"Cape Canaveral SFS, FL, USA"
127,6f94dbe4-1318-4ffb-a53d-351bbdcba733,Falcon 9 Block 5 | Starlink Group 17-49,2026-08-12 04:46:47+00:00,Launch Successful,Falcon 9 Block 5,Starlink,Space Launch Complex 4E,"Vandenberg SFB, CA, USA"
128,36ab6924-bb36-416a-b71d-cda9e88004be,Falcon 9 Block 5 | Starlink Group 10-19,2026-08-11 15:58:26+00:00,Launch Successful,Falcon 9 Block 5,Starlink,Space Launch Complex 40,"Cape Canaveral SFS, FL, USA"


In [33]:
from pathlib import Path

PROCESSED_DIR = Path.cwd().parent / "data" / "processed"
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

launch_path = PROCESSED_DIR / "falcon9_launches_clean.csv"
df_launches_clean.to_csv(launch_path, index=False)

print(f"Clean launch dataset saved → {launch_path.name}")
print(f"Total historical Falcon 9 launches: {len(df_launches_clean)}")

Clean launch dataset saved → falcon9_launches_clean.csv
Total historical Falcon 9 launches: 680


In [34]:
print("=== NOTEBOOK 02 FINAL STATUS ===")
print(f"Historical Falcon 9 launches : {len(df_launches_clean)}")
print(f"Clean Falcon 9 boosters      : {len(df_boosters_clean)}")
print("\nBoth datasets are now ready for analysis in Notebook 03.")

=== NOTEBOOK 02 FINAL STATUS ===
Historical Falcon 9 launches : 680
Clean Falcon 9 boosters      : 86

Both datasets are now ready for analysis in Notebook 03.


## Final Research Notes – Notebook 02

Notebook 02 prepares the Falcon 9 launch and booster datasets for analysis.  
It does **not** yet draw conclusions about reusability, cadence, or capacity. Those conclusions belong to Notebook 03.

### Key numbers to review before moving forward:
1. Total Falcon 9 launch records
2. % with valid booster serial
3. % with valid booster flight number
4. Number of unique boosters
5. Number of boosters with ≥ 2 observed flights
6. Number of launches with calculable turnaround
7. Number of questionable/unknown booster records

These seven numbers will determine whether Launch Library 2 alone is sufficient for the reusability analysis, or whether we need to supplement i

## Research Notes – Key Data Quality Findings (Notebook 02)

### 1. Total Falcon 9 launch records
- Clean historical Falcon 9 launches: **680**
- This is a solid sample size for analysis.

### 2. % with valid booster serial (from launch records)
- **0%**
- The original launch data from Launch Library 2 did not contain usable booster serial numbers.

### 3. % with valid booster flight number (from launch records)
- **0%**
- Flight count information was also missing from the launch records.

### 4. Number of unique boosters
- Collected a separate booster dataset containing **86** clean Falcon 9 boosters.

### 5. Boosters with ≥ 2 flights
- The majority of the 86 boosters have flown multiple times.
- Highest recorded flight count: **B1067 with 36 flights**.

### 6. Launches with calculable turnaround
- **0** from the original launch table (due to missing serial numbers).
- Turnaround analysis will rely on the dedicated booster dataset going forward.

### 7. Questionable / unknown booster records
- “Unknown F9” records were removed.
- Final clean booster count: **86**.

### Key Methodological Lesson
The launch endpoint of Launch Library 2 is strong for cadence and mission classification, but weak for booster-level reuse tracking.  
We therefore collected booster data from the dedicated `/launchers/` endpoint. This is a more reliable source for reusability analysis.

### Ready for Notebook 03
We now have two clean datasets:
- Historical Falcon 9 launches (680 records)
- Falcon 9 boosters (86 records)

These will form the foundation for cadence and reusability analysis.